### PR TITLE
[Algorithm] Update TD3 Example

### PR DESCRIPTION
**Pull request by BY571**
## Description

Updated TD3 script similar to [PPO Update](https://github.com/pytorch/rl/pull/1495), added time logging, more comments,  cleaner structure, and fixes here and there. Running some tests right now to verify performance.
What do you think @vmoens, @albertbou92 @matteobettini, how could we improve the example further?

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: examples/td3/config.yaml [hunk 1]
```diff
@@ -1,47 +1,49 @@
-# Environment
+# task and env
 env:
   name: HalfCheetah-v3
   task: ""
-  exp_name: "HalfCheetah-TD3"
+  exp_name: "HalfCheetah-TD3-ICLR"
```
Comment (vmoens) [inline examples/td3/config.yaml:5]: Can we avoid any specific reference to a submission that we may forget to remove?

File: examples/td3/config.yaml [hunk 2]
```diff
@@ -1,47 +1,49 @@
...
-# Collection
+# collector
 collector:
-  total_frames: 1000000
-  init_random_frames: 10000
+  total_frames: 3000000
+  init_random_frames: 25_000
   init_env_steps: 1000
-  frames_per_batch: 1000
+  frames_per_batch: 1
```
Comment (vmoens) [inline examples/td3/config.yaml:15]: this is going to be super inefficient, do we have to stick to that?
can we obtain the same perf with something faster?

File: examples/td3/td3.py [hunk 3]
```diff
@@ -45,140 +47,156 @@ def main(cfg: "DictConfig"):  # noqa: F821
...
+                # Compute loss
                 loss_td = loss_module(sampled_tensordict)
 
                 actor_loss = loss_td["loss_actor"]
                 q_loss = loss_td["loss_qvalue"]
 
+                # Update critic
                 optimizer_critic.zero_grad()
-                update_actor = j % delayed_updates == 0
+                update_actor = update_counter % delayed_updates == 0
```
Comment (vmoens) [inline examples/td3/td3.py:131]: let's move this upward, under update_counter (or move update_counter down)

File: examples/td3/td3.py [hunk 4]
```diff
@@ -45,140 +47,156 @@ def main(cfg: "DictConfig"):  # noqa: F821
...
-            "train_reward": rewards[-1][1],
-            "collected_frames": collected_frames,
-        }
-        if q_loss is not None:
-            train_log.update(
-                {
-                    "actor_loss": np.mean(actor_losses),
-                    "q_loss": np.mean(q_losses),
-                }
+        training_time = time.time() - training_start
```
Comment (vmoens) [inline examples/td3/td3.py:151]: We should not need that (if we're using wandb or any other logger) but happy to keep it if that makes logging the time spent clearer

**Review by vmoens — APPROVED** [review]
LGTM! Some minor comments

File: examples/td3/td3.py (hunk 4 above)
Comment (BY571) [reply examples/td3/td3.py:155]: I think it gives a better overview of how much time is spent at each big step (sample, update, and test) in the training loop. But I'm also okay with removing it as you said wandb and other logger keep track of timing themselves.

File: examples/td3/config.yaml (hunk 2 above)
Comment (BY571) [reply examples/td3/config.yaml]: Performance might be a bit less stable, I'm testing it right now.

File: examples/td3/td3.py (hunk 4 above)
Comment (vmoens) [reply examples/td3/td3.py:155]: we can keep that no big deal

File: examples/td3/utils.py
Comment (vmoens) [inline examples/td3/utils.py:1]: because we're using gym and not gymnasium, we should decorate calls to GymEnv with `with set_gym_backend("gym")`

Also, the headers are missing

File: torchrl/objectives/td3.py
Comment (vmoens) [inline torchrl/objectives/td3.py:1]: If we're refactoring td3, maybe we can use the changed from #1377?

Comment (BY571) [reply torchrl/objectives/td3.py:1]: I updated the loss like in #1377

Comment (BY571) [reply torchrl/objectives/td3.py:1]: Forgot to push, should be updated now. Also started some tests to check performance. CI test should be also up-to-date

Comment (vmoens) [reply torchrl/objectives/td3.py:1]: Got it
The TD3 tests are failing so we should investigate what is going on

File: test/test_cost.py [hunk 5]
```diff
@@ -2357,8 +2357,10 @@ def test_td3_notensordict(self, observation_key, reward_key, done_key):
             loss_val_td = loss(td)
             torch.manual_seed(0)
             loss_val = loss(**kwargs)
-            for i, key in enumerate(loss_val_td.keys()):
-                torch.testing.assert_close(loss_val_td.get(key), loss_val[i])
+            for i in loss_val:
+                assert i in loss_val_td.values(), f"{i} not in {loss_val_td.values()}"
+            # for i, key in enumerate(loss_val_td.keys()):
+            # torch.testing.assert_close(loss_val_td.get(key), loss_val[i])
```
Comment (BY571) [inline test/test_cost.py:2363]: This is dangerous as keys in the tensordict get ordered by name but output tuple loss_val doesn't. So for now im just checking if all values in the loss_val tuple are also in the loss_val_td.


```
# actor metadata
metadata = {
            "state_action_value_actor": state_action_value_actor.mean().detach(),
        }
# value metadata
        metadata = {
            "td_error": td_error,
            "next_state_value": next_target_qvalue.mean().detach(),
            "pred_value": current_qvalue.mean().detach(),
            "target_value": target_value.mean().detach(),
        }
# out tensordict
        td_out = TensorDict(
            source={
                "loss_actor": loss_actor,
                "loss_qvalue": loss_qval,
                **metadata_actor,
                **metadata_value,
            },
            batch_size=[],
        )
```
loss_vals will be in that order `(loss_actor, loss_qvalue, state_action_value_actor, next_state_value, pred_value, target_value)`
However, as the items are getting ordered in the TD by the keys the output tensordict has actually this order:
 `(loss_actor, loss_qvalue, next_state_value, pred_value, state_action_value_actor, target_value)`

Comment (vmoens) [reply test/test_cost.py:2423]: dispatch returns the keys in the order of out_keys.
So this is predictable, we can just do
```python
            for i, key in enumerate(loss.out_keys):
                torch.testing.assert_close(loss_val_td.get(key), loss_val[i])
```
does that solve the problem?

**Review by vmoens — APPROVED** [review]
LGTM
let's wait for the tests to pass!
